File: internal/orca/orca.go
package orca

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"

	"github.com/ryusudol/plexus/internal/jsonx"
	"github.com/ryusudol/plexus/internal/paths"
)

type Pane struct {
	SessionID string
	Cwd       string
	TabID     string
}

type Focus struct {
	Cwd       string
	SessionID string
}

func DataFiles() []string {
	if env := os.Getenv("ORCA_DATA_FILE"); env != "" {
		return []string{env}
	}
	root := filepath.Join(paths.Home(), "Library/Application Support/orca/profiles")
	entries, err := os.ReadDir(root)
	if err != nil {
		return nil
	}
	out := []string{}
	for _, e := range entries {
		candidate := filepath.Join(root, e.Name(), "orca-data.json")
		if _, err := os.Stat(candidate); err == nil {
			out = append(out, candidate)
		}
	}
	return out
}

func cwdFromWorktreeID(key string) string {
	idx := strings.Index(key, "::")
	if idx < 0 {
		return ""
	}
	return key[idx+2:]
}

func sessionIDFromWorkspace(ws map[string]any) string {
	worktreeID := jsonx.Str(ws["activeWorktreeId"])
	tabID := jsonx.Str(ws["activeTabId"])
	if tabID == "" && worktreeID != "" {
		if by := jsonx.AsMap(ws["activeTabIdByWorktree"]); by != nil {
			tabID = jsonx.Str(by[worktreeID])
		}
	}
	if tabID == "" {
		return ""
	}
	panes := jsonx.AsMap(ws["sleepingAgentSessionsByPaneKey"])
	layouts := jsonx.AsMap(ws["terminalLayoutsByTabId"])
	var leafID string
	if layouts != nil {
		if layout := jsonx.AsMap(layouts[tabID]); layout != nil {
			leafID = jsonx.Str(layout["activeLeafId"])
		}
	}
	if leafID != "" && panes != nil {
		if pane := jsonx.AsMap(panes[tabID+":"+leafID]); pane != nil {
			if ps := jsonx.AsMap(pane["providerSession"]); ps != nil {
				if id := jsonx.Str(ps["id"]); id != "" {
					return id
				}
			}
		}
	}
	if panes == nil {
		return ""
	}
	for _, raw := range panes {
		pane := jsonx.AsMap(raw)
		if pane == nil || jsonx.Str(pane["tabId"]) != tabID {
			continue
		}
		if ps := jsonx.AsMap(pane["providerSession"]); ps != nil {
			if id := jsonx.Str(ps["id"]); id != "" {
				return id
			}
		}
	}
	return ""
}

func ReadLivePanes(file string) (map[string]Pane, bool) {
	files := DataFiles()
	if file != "" {
		files = []string{file}
	}
	if len(files) == 0 {
		return nil, false
	}
	panes := map[string]Pane{}
	saw := false
	for _, item := range files {
		b, err := os.ReadFile(item)
		if err != nil {
			continue
		}
		var obj map[string]any
		if json.Unmarshal(b, &obj) != nil {
			continue
		}
		saw = true
		ws := jsonx.AsMap(obj["workspaceSession"])
		if ws == nil {
			ws = map[string]any{}
		}
		openTabs := map[string]bool{}
		if tabsBy := jsonx.AsMap(ws["tabsByWorktree"]); tabsBy != nil {
			for _, tabs := range tabsBy {
				for _, tab := range jsonx.Slice(tabs) {
					if tm := jsonx.AsMap(tab); tm != nil {
						if id := jsonx.Str(tm["id"]); id != "" {
							openTabs[id] = true
						}
					}
				}
			}
		}
		sleeping := jsonx.AsMap(ws["sleepingAgentSessionsByPaneKey"])
		if sleeping == nil {
			continue
		}
		for _, raw := range sleeping {
			pane := jsonx.AsMap(raw)
			if pane == nil {
				continue
			}
			ps := jsonx.AsMap(pane["providerSession"])
			id := jsonx.Str(ps["id"])
			if id == "" {
				continue
			}
			tabID := jsonx.Str(pane["tabId"])
			if tabID != "" && len(openTabs) > 0 && !openTabs[tabID] {
				continue
			}
			cwd := cwdFromWorktreeID(jsonx.Str(pane["worktreeId"]))
			prev, ok := panes[id]
			if !ok || (cwd != "" && prev.Cwd == "") {
				panes[id] = Pane{SessionID: id, Cwd: cwd, TabID: tabID}
			}
		}
	}
	if !saw {
		return nil, false
	}
	return panes, true
}

func ReadFocus(file string) *Focus {
	files := DataFiles()
	if file != "" {
		files = []string{file}
	}
	var best *Focus
	var bestM int64
	for _, item := range files {
		st, err := os.Stat(item)
		if err != nil {
			continue
		}
		b, err := os.ReadFile(item)
		if err != nil {
			continue
		}
		var obj map[string]any
		if json.Unmarshal(b, &obj) != nil {
			continue
		}
		ws := jsonx.AsMap(obj["workspaceSession"])
		if ws == nil {
			ws = map[string]any{}
		}
		worktreeID := jsonx.Str(ws["activeWorktreeId"])
		if worktreeID == "" {
			if ui := jsonx.AsMap(obj["ui"]); ui != nil {
				worktreeID = jsonx.Str(ui["lastActiveWorktreeId"])
			}
		}
		cwd := cwdFromWorktreeID(worktreeID)
		sessionID := sessionIDFromWorkspace(ws)
		m := st.ModTime().UnixMilli()
		if (cwd != "" || sessionID != "") && m >= bestM {
			best = &Focus{Cwd: cwd, SessionID: sessionID}
			bestM = m
		}
	}
	return best
}

func ReadActiveCwd(file string) string {
	if f := ReadFocus(file); f != nil {
		return f.Cwd
	}
	return ""
}

File: server/orca.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import type { OrcaFocus, OrcaPane } from "../lib/types.ts";

type OrcaPaneRecord = {
  providerSession?: { id?: string };
  tabId?: string;
  worktreeId?: string;
};

type OrcaWorkspace = {
  tabsByWorktree?: Record<string, Array<{ id?: string }>>;
  sleepingAgentSessionsByPaneKey?: Record<string, OrcaPaneRecord>;
  activeWorktreeId?: string;
  activeTabId?: string;
  activeTabIdByWorktree?: Record<string, string>;
  terminalLayoutsByTabId?: Record<string, { activeLeafId?: string }>;
};

type OrcaFile = {
  workspaceSession?: OrcaWorkspace;
  ui?: { lastActiveWorktreeId?: string };
};

export function orcaDataFiles(): string[] {
  if (process.env.ORCA_DATA_FILE) return [process.env.ORCA_DATA_FILE];
  const root = path.join(os.homedir(), "Library/Application Support/orca/profiles");
  const files: string[] = [];
  try {
    for (const name of fs.readdirSync(root)) {
      const candidate = path.join(root, name, "orca-data.json");
      if (fs.existsSync(candidate)) files.push(candidate);
    }
  } catch {
    // none
  }
  return files;
}

function cwdFromWorktreeId(key: unknown): string | null {
  if (typeof key !== "string") return null;
  const idx = key.indexOf("::");
  if (idx < 0) return null;
  return key.slice(idx + 2) || null;
}

function sessionIdFromOrcaWorkspace(ws: OrcaWorkspace = {}): string | null {
  const worktreeId = ws.activeWorktreeId || "";
  const tabId =
    ws.activeTabId ||
    (worktreeId && ws.activeTabIdByWorktree && ws.activeTabIdByWorktree[worktreeId]) ||
    null;
  if (!tabId) return null;
  const panes = ws.sleepingAgentSessionsByPaneKey || {};
  const layout = (ws.terminalLayoutsByTabId || {})[tabId] || {};
  const leafId = layout.activeLeafId;
  if (leafId) {
    const pane = panes[`${tabId}:${leafId}`];
    const id = pane?.providerSession?.id;
    if (typeof id === "string" && id) return id;
  }
  for (const pane of Object.values(panes)) {
    if (pane?.tabId === tabId && typeof pane?.providerSession?.id === "string") {
      return pane.providerSession.id;
    }
  }
  return null;
}

/**
 * Grok sessions currently attached to an open Orca tab.
 * Returns null when Orca state is missing so callers can fall back.
 */
export function readOrcaLivePanes(file?: string): Map<string, OrcaPane> | null {
  const files = file ? [file] : orcaDataFiles();
  if (!files.length) return null;
  const panes = new Map<string, OrcaPane>();
  let saw = false;
  for (const item of files) {
    try {
      const obj = JSON.parse(fs.readFileSync(item, "utf8")) as OrcaFile;
      const ws = obj?.workspaceSession || {};
      saw = true;
      const openTabs = new Set<string>();
      for (const tabs of Object.values(ws.tabsByWorktree || {})) {
        if (!Array.isArray(tabs)) continue;
        for (const tab of tabs) {
          if (tab?.id) openTabs.add(tab.id);
        }
      }
      for (const pane of Object.values(ws.sleepingAgentSessionsByPaneKey || {})) {
        const id = pane?.providerSession?.id;
        if (typeof id !== "string" || !id) continue;
        if (pane?.tabId && openTabs.size && !openTabs.has(pane.tabId)) continue;
        const cwd = cwdFromWorktreeId(pane?.worktreeId);
        const prev = panes.get(id);
        if (!prev || (cwd && !prev.cwd)) panes.set(id, { sessionId: id, cwd, tabId: pane?.tabId || null });
      }
    } catch {
      // skip unreadable profiles
    }
  }
  return saw ? panes : null;
}

export function readOrcaFocus(file?: string): OrcaFocus | null {
  const files = file ? [file] : orcaDataFiles();
  let best: OrcaFocus | null = null;
  let bestM = 0;
  for (const item of files) {
    try {
      const m = fs.statSync(item).mtimeMs;
      const obj = JSON.parse(fs.readFileSync(item, "utf8")) as OrcaFile;
      const ws = obj?.workspaceSession || {};
      const worktreeId = ws.activeWorktreeId || obj?.ui?.lastActiveWorktreeId || "";
      const cwd = cwdFromWorktreeId(worktreeId);
      const sessionId = sessionIdFromOrcaWorkspace(ws);
      if ((cwd || sessionId) && m >= bestM) {
        best = { cwd, sessionId };
        bestM = m;
      }
    } catch {
      // skip
    }
  }
  return best;
}

export function readOrcaActiveCwd(file?: string): string | null {
  return readOrcaFocus(file)?.cwd || null;
}
